refactor!: modify C implementation to accept `double` instead of `int32` in `math/base/special/fibonacci`

gunjjoshi opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: na
  - task: lint_repl_help status: na
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: na
  - task: lint_javascript_tests status: na
  - task: lint_javascript_benchmarks status: na
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: passed
  - task: lint_c_examples status: passed
  - task: lint_c_benchmarks status: passed
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: na
  - task: lint_typescript_tests status: na
  - task: lint_license_headers status: passed ---

Resolves None.

## Description

> What is the purpose of this pull request?

This pull request:

-   modifies the C implementation to accept `double` instead of `int32` in [`math/base/special/fibonacci`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/fibonacci)

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves None.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

gunjjoshi committed:
refactor: modify C implementation to accept double

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/fibonacci/src/main.c:
```diff
@@ -108,12 +110,12 @@ static const int64_t fibonacci_value[ 79 ] = {
 		return 0.0 / 0.0; // NaN
 	}
-	return fibonacci_value[ n ];
+	return fibonacci_value[ (int8_t)n ];
```
As `n` won't be greater than `78`, I have used `int8_t` here.

kgryte replied:
This should be fine; although, it may be better to cast to `size_t`, which is likely (?) what a C compiler will do after casting to `int8_t`. We can skip that intermediate cast step.

gunjjoshi replied:
Got it, I've updated it now to use `size_t` instead. Thanks!

kgryte commented:
cc @anandkaranubc as a heads up, this is something Gunj and I discussed. Namely, to update a few function signatures to ensure that we have consistent results between C and JS. Otherwise, if things are inconsistent, it makes scaffolding much more difficult and results in likely unexpected divergences between C and JS in userland.

gunjjoshi committed:
refactor: replace int8_t by size_t

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte reviewed (approved):
LGTM

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/fibonacci/lib/main.js:
```diff
@@ -77,8 +77,7 @@ var FIBONACCI = require( './fibonacci.json' );
 		isnan( n ) ||
-		isInteger( n ) === false ||
-		n < 0 ||
+		!isNonnegativeInteger( n ) ||
```
@kgryte I was thinking here, regarding what do we prefer? We can have two ways:
```
isNonnegativeInteger( n ) === false
```
and
```
!isNonnegativeInteger
```
Do we have a preference here? Or both are fine?

kgryte replied:
Depends on the case. Here, either is fine.
